scio-bigquery/src/main/scala/com/spotify/scio/bigquery/nio/BigQueryIO.scala

```diff
@@ -0,0 +1,286 @@
+   *
+   * The source (table) specified in the annotation will be used
+   *
+   * {{{
+   * @BigQueryType.fromTable("publicdata:samples.gsod")
+   * class Row
+   * }}}
+   *
+   */
+  def apply[T <: HasAnnotation with HasTable : ClassTag : TypeTag]: ScioIO.Aux[T, Unit, Nothing] = {
+    val bqt = BigQueryType[T]
+    val _table: String = bqt.table.get
+    table(_table)
+  }
+
+  /**
+   * Get a typed SCollection for a BigQuery SELECT query
+   *
+   * Note that `T` must be annotated with
+   * [[com.spotify.scio.bigquery.types.BigQueryType.fromQuery BigQueryType.fromQuery]]
+   *
+   * Both [[https://cloud.google.com/bigquery/docs/reference/legacy-sql Legacy SQL]] and
+   * [[https://cloud.google.com/bigquery/docs/reference/standard-sql/ Standard SQL]] dialects are
+   * supported. By default the query dialect will be automatically detected. To override this
+   * behavior, start the query string with `#legacysql` or `#standardsql`.
+   */
+  def apply[T <: HasAnnotation with HasQuery : ClassTag : TypeTag](implicit dummy: DummyImplicit): ScioIO.ReadOnly[T, Unit] = {
+    val bqt = BigQueryType[T]
+    val _query = bqt.query.get
+    query(_query)
+  }
+
+  /**
+   * Get a typed SCollection for a BigQuery SELECT query
+   *
+   * Both [[https://cloud.google.com/bigquery/docs/reference/legacy-sql Legacy SQL]] and
+   * [[https://cloud.google.com/bigquery/docs/reference/standard-sql/ Standard SQL]] dialects are
+   * supported. By default the query dialect will be automatically detected. To override this
+   * behavior, start the query string with `#legacysql` or `#standardsql`.
+   */
+  def query[T <: HasAnnotation : ClassTag : TypeTag](query: String): ScioIO.ReadOnly[T, Unit] =
+    new ScioIO[T] {
+      type ReadP = Unit
+      type WriteP = Nothing // ReadOnly
+
+      def id: String = query
+      def read(sc: ScioContext, params: ReadP): SCollection[T] = {
+        @inline def typedRead(sc: ScioContext) = Reads.avroBigQueryRead[T](sc)
+        Reads.bqReadQuery(sc)(typedRead(sc), query)
+      }
+
+      def write(data: SCollection[T], params: WriteP): Future[Tap[T]] = ???
+      def tap(read: ReadP): Tap[T] = ???
+    }
+
+  /**
+   * Get a typed SCollection for a BigQuery table.
+   */
+  def table[T <: HasAnnotation : ClassTag : TypeTag](table: String): ScioIO.Aux[T, Unit, Nothing] =
+    new ScioIO[T] {
+      type ReadP = Unit
+      type WriteP = Nothing // TODO
+
+      def id: String = table
+
+      def read(sc: ScioContext, params: ReadP): SCollection[T] = {
+        @inline def typedRead(sc: ScioContext) = Reads.avroBigQueryRead[T](sc)
+        Reads.bqReadTable(sc)(typedRead(sc), bqio.BigQueryHelpers.parseTableSpec(id))
+      }
+
+      def write(data: SCollection[T], params: WriteP): Future[Tap[T]] = ???
+      def tap(read: ReadP): Tap[T] = ???
+    }
+
+  private[scio] def dynamic[T <: HasAnnotation : ClassTag : TypeTag](
+    newSource: String
+  ): ScioIO.ReadOnly[T, Unit] = {
```

Comment:
This method only exists to support `BigQueryScioContext.typedBigQuery` which is why it's private.